### DISCUSSION TITLE
Can OMC generate only the logical code in C language from MO (Model) files, without including any solver code for simulation?

### DISCUSSION THREAD

**Discussion by pily1** [OpenModelica/OpenModelica#12599] [Q&A] [answered]
Can OMC generate only the logical code in C language from MO (Model) files, without including any solver code for simulation?
if it can,how to use it?
For example, if the mo file code look like this:

```modelica
model MyModel
  input Integer x;
  input Integer y;
  output Integer z;
algorithm
  z :=x+y;
end MyModel;
```

the translated to C language code look like this:

```C
struct MyModelSt
{
      int x;
      int y;
      int z;
};

int MyModel_eqFunction(MyModelSt input)
{
      input.z = input.x + input.y;
      return input.z;
}
 ```

**Comment by AnHeuermann** [comment]
Not sure what exactly the "logical code" you are referring to should be. But if you want to get a C representation of the equation system in ODE form the answer is *yes*.

Have a look at the generated C files for your `MyModel.mo`. In `MyModel.c` you'll find a C function for each equation:

```C
/*
equation index: 2
type: ALGORITHM

  z := pre(z);
  z := x + y;
*/
void MyModel_eqFunction_2(DATA *data, threadData_t *threadData)
{
  TRACE_PUSH
  const int equationIndexes[2] = {1,2};
  (data->localData[0]->integerVars[2] /* z DISCRETE */) = (data->simulationInfo->integerVarsPre[2] /* z DISCRETE */);

  (data->localData[0]->integerVars[2] /* z DISCRETE */) = (data->localData[0]->integerVars[0] /* x variable */) + (data->localData[0]->integerVars[1] /* y variable */);
  TRACE_POP
}
```

And you have a few C functions to evaluate all of the right hand side of the ODE and DAE and so on.
If an equation systems needs to be evaluated naturally one needs a solver to so. For example a Newton solver to solve some non-linear equation system.
Before calling the right hand side you'll need to do few additional steps: Initialize solvers and memory, read parameters and start values, update inputs, update external objects / functions.

Now to use these C files for anything serious there are two huge caveats:

1. OpenModelica makes no promises at all that this very internal structure won't change.
    In fact as soon as the Backend is updated the next logical step is to update the C code. So it's very likely that this will change completely and it could be anything from basically the same C files to C++ or Rust files or no files at all and instead some LLVM stuff is used to directly compile the simulation executable.
2. You'll need to link the OpenModelica C runtime to this code to get access to the used data structures.
    Of yours you can do a minimal version of it and only define the types. But that effort is lost whenever issue 1 hits.

**Reply by AnHeuermann** [reply]
Of course you could write your own code generation targeting your desired format. But that's a very long endeavor and not easy at all. And whenever OpenModelica is updated you'll need to update your code generation.
At that point you might want to use a tool that can do index reduction and and BLT transformation and then do whatever you want to do with the resulting set of equations.  Maybe have a look at https://github.com/SciML/ModelingToolkit.jl

**Reply by pily1** [reply]
I want to write my own code generation targeting my desired format, which parts of the code in OpenModelica that are responsible for generating C language code？

**Reply by AnHeuermann** [reply, marked as answer]
Of course you can do that. OpenModelica is open-source and there exits some tools that build on top of OpenModelica.

You can start with [OMCompiler/Compiler/Template](https://github.com/OpenModelica/OpenModelica/tree/master/OMCompiler/Compiler/Template). You'll find the different code generation templates in that directory. You want something similar to CodegenC.tpl.
The code generation generate strings for the data provided in the [SimCode](https://github.com/OpenModelica/OpenModelica/blob/master/OMCompiler/Compiler/SimCode/SimCode.mo#L107) structure. The languages you'll need to get used to are MetaModelica and Susan. There isn't a lot of documentation around. The best way I know is to read the code and try to understand and maybe look at the publications for both languages.

But to give you a fair warning: This would be an endeavor spanning the next few months and it is very unlikely that a contribution with this code generation would be accepted in OpenModelica.

**Reply by pily1** [reply]
think you very much,I will try it

**Comment by casella** [comment]
I would suggest using FMI-ME for this. The interface is well-defined and stable. Our C code may change in the future.

**Reply by pily1** [reply]
the fmu has many useless code for me,I just want to get a C representation of the equation system in ODE form
